diff --git a/Rendering/Core/vtkAssembly.cxx b/Rendering/Core/vtkAssembly.cxx
index 8961c26..5d8dd0f 100644
--- a/Rendering/Core/vtkAssembly.cxx
+++ b/Rendering/Core/vtkAssembly.cxx
@@ -369,10 +369,12 @@ double *vtkAssembly::GetBounds()
     vtkProp3D* prop3D = static_cast<vtkProp3D *>(path->GetLastNode()->GetViewProp());
     if (prop3D->GetVisibility() && prop3D->GetUseBounds())
       {
-      propVisible = 1;
       prop3D->PokeMatrix(path->GetLastNode()->GetMatrix());
       double* bounds = prop3D->GetBounds();
       prop3D->PokeMatrix(NULL);
+      if ( !vtkMath::AreBoundsInitialized(bounds) )
+        continue;
+      propVisible = 1;
 
       double bbox[24];
       // fill out vertices of a bounding box
