diff --git a/Rendering/Annotation/vtkAxisActor.cxx b/Rendering/Annotation/vtkAxisActor.cxx
index 01101a5..56d95b8 100644
--- a/Rendering/Annotation/vtkAxisActor.cxx
+++ b/Rendering/Annotation/vtkAxisActor.cxx
@@ -69,6 +69,7 @@ vtkAxisActor::vtkAxisActor()
   this->TitleTextProperty = vtkTextProperty::New();
   this->TitleTextProperty->SetColor(0.,0.,0.);
   this->TitleTextProperty->SetFontFamilyToArial();
+  this->TitleTextProperty->SetFontSize(14);
 
   this->TitleVector = vtkVectorText::New();
   this->TitleMapper = vtkPolyDataMapper::New();
@@ -88,6 +89,7 @@ vtkAxisActor::vtkAxisActor()
   this->LabelTextProperty = vtkTextProperty::New();
   this->LabelTextProperty->SetColor(0.,0.,0.);
   this->LabelTextProperty->SetFontFamilyToArial();
+  this->LabelTextProperty->SetFontSize(12);
 
   this->AxisLines = vtkPolyData::New();
   this->AxisLinesMapper = vtkPolyDataMapper::New();
@@ -682,6 +684,7 @@ vtkAxisActor::BuildLabels(vtkViewport *viewport, bool force)
   for (int i = 0; i < this->NumberOfLabelsBuilt; i++)
     {
     this->LabelActors[i]->SetCamera(this->Camera);
+    this->LabelActors[i]->SetProperty(this->GetProperty());
     this->LabelActors[i]->GetProperty()->SetColor(this->LabelTextProperty->GetColor());
     this->LabelActors[i]->SetOrientation(0., 0., this->LabelTextProperty->GetOrientation());
 
@@ -795,7 +798,7 @@ vtkAxisActor::BuildLabels2D(vtkViewport *viewport, bool force)
     {
     this->LabelActors2D[i]->GetProperty()->SetColor(this->LabelTextProperty->GetColor());
     this->LabelActors2D[i]->GetProperty()->SetOpacity(1);
-    this->LabelActors2D[i]->GetTextProperty()->SetFontSize(14);
+    this->LabelActors2D[i]->GetTextProperty()->SetFontSize(this->LabelTextProperty->GetFontSize());
     this->LabelActors2D[i]->GetTextProperty()->SetVerticalJustificationToBottom();
     this->LabelActors2D[i]->GetTextProperty()->SetJustificationToLeft();
     }
@@ -985,7 +988,7 @@ vtkAxisActor::BuildTitle2D(vtkViewport *viewport, bool force)
   this->TitleActor2D->SetInput( this->TitleVector->GetText() );
   this->TitleActor2D->GetProperty()->SetColor( this->TitleTextProperty->GetColor() );
   this->TitleActor2D->GetProperty()->SetOpacity(1);
-  this->TitleActor2D->GetTextProperty()->SetFontSize(18);
+  this->TitleActor2D->GetTextProperty()->SetFontSize( this->TitleTextProperty->GetFontSize() );
   this->TitleActor2D->GetTextProperty()->SetVerticalJustificationToCentered();
   this->TitleActor2D->GetTextProperty()->SetJustificationToCentered();
 
diff --git a/Rendering/Core/vtkAssembly.cxx b/Rendering/Core/vtkAssembly.cxx
index 7549196..f85b6c1 100644
--- a/Rendering/Core/vtkAssembly.cxx
+++ b/Rendering/Core/vtkAssembly.cxx
@@ -388,10 +388,12 @@ double *vtkAssembly::GetBounds()
     prop3D = static_cast<vtkProp3D *>(path->GetLastNode()->GetViewProp());
     if ( prop3D->GetVisibility() && prop3D->GetUseBounds() )
       {
-      propVisible = 1;
       prop3D->PokeMatrix(path->GetLastNode()->GetMatrix());
       bounds = prop3D->GetBounds();
       prop3D->PokeMatrix(NULL);
+      if ( !vtkMath::AreBoundsInitialized(bounds) )
+        continue;
+      propVisible = 1;
 
       // fill out vertices of a bounding box
       bbox[ 0] = bounds[1]; bbox[ 1] = bounds[3]; bbox[ 2] = bounds[5];
