Index: Rendering/Core/vtkClipPlanesPainter.cxx
===================================================================
--- Rendering/Core/vtkClipPlanesPainter.cxx	(revision 1808)
+++ Rendering/Core/vtkClipPlanesPainter.cxx	(working copy)
@@ -19,7 +19,6 @@
 #include "vtkInformationObjectBaseKey.h"
 #include "vtkObjectFactory.h"
 #include "vtkPlaneCollection.h"
-#include "vtkBoundingBox.h"
 
 // Return NULL if no override is supplied.
 vtkAbstractObjectFactoryNewMacro(vtkClipPlanesPainter)
@@ -65,32 +64,3 @@
     os << " (none)" << endl;
     }
 }
-//-----------------------------------------------------------------------------
-void vtkClipPlanesPainter::UpdateBounds(double bounds[6])
-{
-  if(!vtkBoundingBox::IsValid(bounds))
-    {
-    return;
-    }
-  vtkPlaneCollection* planes =this->ClippingPlanes;
-  if(planes)
-    {
-    int numPlanes = planes->GetNumberOfItems();
-    for(int i=0; i<numPlanes; i++)
-      {
-      vtkPlane *plane = planes->GetItem(i);
-      if(plane)
-        {
-        double n[3],p[3];
-        plane->GetNormal(n);
-        plane->GetOrigin(p);
-        vtkBoundingBox bb(bounds);
-        if(bb.IntersectPlane(p,n))
-          {
-          bb.GetBounds(bounds);
-          }
-        }
-      }
-    }
-
-}
Index: Rendering/Core/vtkClipPlanesPainter.h
===================================================================
--- Rendering/Core/vtkClipPlanesPainter.h	(revision 1808)
+++ Rendering/Core/vtkClipPlanesPainter.h	(working copy)
@@ -35,10 +35,6 @@
   // Get/Set the vtkPlaneCollection which specifies the clipping planes.
   static vtkInformationObjectBaseKey* CLIPPING_PLANES();
 
-  // Description:
-  // Update the bounds based on the clip planes
-  virtual void UpdateBounds(double bounds[6]);
-
 protected:
   vtkClipPlanesPainter();
   ~vtkClipPlanesPainter();
