Index: Rendering/FreeType/vtkTextActor3D.cxx
===================================================================
--- Rendering/FreeType/vtkTextActor3D.cxx	(revision 1700)
+++ Rendering/FreeType/vtkTextActor3D.cxx	(working copy)
@@ -257,6 +257,8 @@
       return 0;
       }
 
+    this->ImageData->Modified();
+
     // Associate the image data (should be up to date now) to the image actor
     if (this->ImageActor)
       {
